archive/imo/imo1960_q1.lean

```diff
@@ -0,0 +1,200 @@
+sum_of_digit_squares expression.
+These are basically "hints" to speed up norm_num.
+-/
+
+def sum_of_digit_squares (n : ℕ) := sum_of_squares (nat.digits 10 n)
+
+lemma sods_zero : sum_of_digit_squares 0 = 0 :=
+by norm_num [sum_of_digit_squares, sum_of_squares]
+
+lemma calc_sods (n : ℕ) (h1 : 0 < n) :
+sum_of_digit_squares n = (n % 10) ^ 2 + (sum_of_digit_squares (n / 10)) :=
+begin
+  unfold sum_of_digit_squares,
+  rw [nat.digits, nat.digits_aux_def _ _ _ h1],
+  unfold sum_of_squares,
+  simp [pow_two],
+end
+
+/-
+This part is doing an exhaustive search.
+-/
+
+def fails_sum (c : ℕ) := c ≠ sum_of_digit_squares (c*11)
+
+def multiples_of_11 (n : ℕ) (h1: problem_predicate n) :
+∃ c : ℕ, ¬ fails_sum c ∧ 9 < c ∧ c < 91 ∧ n = c * 11 :=
+begin
+  have h2: ∃ c : ℕ, n = c * 11, from exists_eq_mul_left_of_dvd h1.right.left,
+  cases h2 with c h3,
+  use c,
+  split,
+  { have h4: c = (c * 11) / 11, by simp,
+    have h5: c = n / 11, from h3.symm ▸ h4,
+    unfold fails_sum,
+    rw [h3.symm, h5],
+    simp,
+    exact h1.right.right },
+  split,
+  { have h6: n ≥ 100, from ge_100 n h1,
+    linarith },
+  have h7: n < 1000, from lt_1000 n h1,
+  split,
+  linarith,
+  exact h3,
+end
+
+lemma step_helper (b c : ℕ) (h : b < c) : b + 1 = c ∨ b + 1 < c := eq_or_lt_of_le h
+
+lemma iterative_step (c bound : ℕ) (h1 : fails_sum (bound + 1)) (h2 : fails_sum c ∨ bound < c) :
+fails_sum c ∨ bound + 1 < c :=
+begin
+  cases h2,
+  { left,
+    exact h2 },
+  have h3 : bound + 1 = c ∨ bound + 1 < c, from step_helper bound c h2,
+  cases h3,
+  { left,
+    exact h3 ▸ h1 },
+  right,
+  exact h3,
+end
+
+lemma low_search (c : ℕ) (lower_bound: 9 < c) : fails_sum c ∨ 49 < c :=
+begin
+  iterate 40 {
+    refine iterative_step _ _ _ _,
+    norm_num [fails_sum, calc_sods, sods_zero],
+  },
+  right,
+  exact lower_bound,
+end
+
+lemma mid_search (c : ℕ) (lower_bound : 50 < c) : fails_sum c ∨ 72 < c :=
+begin
+  iterate 22 {
+    refine iterative_step _ _ _ _,
+    norm_num [fails_sum, calc_sods, sods_zero],
+  },
+  right,
+  exact lower_bound,
+end
+
+lemma high_search (c : ℕ) (lower_bound : 73 < c) : fails_sum c ∨ 90 < c :=
+begin
+  iterate 17 {
+    refine iterative_step _ _ _ _,
+    norm_num [fails_sum, calc_sods, sods_zero],
+  },
+  right,
+  exact lower_bound,
+end
+
+/-
+Now we just need to combine the results from the `search` lemmas.
+-/
+
+lemma right_direction (n : ℕ) : problem_predicate n → solution_predicate n :=
+begin
+  intro ppn,
+  have h1 : ∃ c : ℕ, ¬ fails_sum c ∧ 9 < c ∧ c < 91 ∧ n = c * 11, from multiples_of_11 n ppn,
+  cases h1 with c h2,
+  have h3 : fails_sum c ∨ 49 < c, from low_search c h2.right.left,
+  cases h3,
+  exact absurd h3 h2.left,
+  have h4 : 50 = c ∨ 50 < c, from step_helper 49 c h3,
+  unfold solution_predicate,
+  have h5 : n = c * 11, from h2.right.right.right,
+  cases h4,
+  { left,
+    linarith },
+  have h6 : fails_sum c ∨ 72 < c, from mid_search c h4,
+  cases h6,
+  exact absurd h6 h2.left,
```

Comment:
Can you add braces throughout this proof for subgoals?  It would help to see, eg, what this `exact` is answering.

Reply:
Hmm, I thought I did. All the multi-line proofs of subgoals have braces around them. Where else do you want braces?

Reply:
We generally put them even for single-line proofs of subgoals!  See eg
https://leanprover-community.github.io/contribute/style.html

Reply:
Okay, I think I get it. I added braces around every line that is proving a subgoal.